lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs

```diff
@@ -0,0 +1,1009 @@
+// limitations under the License.
+
+use crate::errors::invalid_data_error;
+use crate::{vchan, Payload};
+use bitflags::bitflags;
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use num_traits::FromPrimitive;
+use rdp::core::{mcs, tpkt};
+use rdp::model::data::Message;
+use rdp::model::error::*;
+use rdp::try_let;
+use std::collections::HashMap;
+use std::io::{Read, Write};
+
+pub const CHANNEL_NAME: &str = "cliprdr";
+
+/// Client implements a client for the clipboard virtual channel
+/// (CLIPRDR) extension, as defined in:
+/// https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpeclip/fb9b7e0b-6db4-41c2-b83c-f889c1ee7688
+pub struct Client {
+    clipboard: HashMap<u32, Vec<u8>>,
+    on_remote_copy: Box<dyn Fn(Vec<u8>)>,
+}
+
+impl Default for Client {
+    fn default() -> Self {
+        Self::new(Box::new(|_| {}))
+    }
+}
+
+impl Client {
+    pub fn new(on_remote_copy: Box<dyn Fn(Vec<u8>)>) -> Self {
+        Client {
+            clipboard: HashMap::new(),
+            on_remote_copy,
+        }
+    }
+
+    pub fn read<S: Read + Write>(
+        &mut self,
+        payload: tpkt::Payload,
+        mcs: &mut mcs::Client<S>,
+    ) -> RdpResult<()> {
+        let mut payload = try_let!(tpkt::Payload::Raw, payload)?;
+        let _pdu_header = vchan::ChannelPDUHeader::decode(&mut payload)?;
+        let header = ClipboardPDUHeader::decode(&mut payload)?;
+        debug!("received {:?}", header.msg_type);
+
+        let resp = match header.msg_type {
+            ClipboardPDUType::CB_CLIP_CAPS => self.handle_server_caps(&mut payload)?,
+            ClipboardPDUType::CB_MONITOR_READY => self.handle_monitor_ready(&mut payload)?,
+            ClipboardPDUType::CB_FORMAT_LIST => {
+                self.handle_format_list(&mut payload, header.data_len)?
+            }
+            ClipboardPDUType::CB_FORMAT_LIST_RESPONSE => {
+                self.handle_format_list_response(header.msg_flags)?
+            }
+            ClipboardPDUType::CB_FORMAT_DATA_REQUEST => {
+                self.handle_format_data_request(&mut payload)?
+            }
+            ClipboardPDUType::CB_FORMAT_DATA_RESPONSE => {
+                if header
+                    .msg_flags
+                    .contains(ClipboardHeaderFlags::CB_RESPONSE_OK)
+                {
+                    self.handle_format_data_response(&mut payload, header.data_len)?
+                } else {
+                    warn!("RDP server failed to process format data request");
+                    None
+                }
+            }
+            _ => {
+                warn!(
+                    "CLIPRDR message {:?} not implemented, ignoring",
+                    header.msg_type
+                );
+                None
+            }
+        };
+
+        if let Some(resp) = resp {
+            let chan = &CHANNEL_NAME.to_string();
+            for msg in resp {
+                mcs.write(chan, msg)?;
+            }
+        }
```

Comment:
Bordering on a nitpick, but could this be flattened? `Option<Vec<Vec<u8>>>` (I think representing an optional list of replies?) might be the equivalent to just an empty vs non-empty `Vec<Vec<u8>>`.

The handler functions could return `Ok(vec![])` instead of `Ok(None)` for empty replies.

Reply:
Oh, very good point. Will fix.

Reply:
One thing to note if you are unaware is that empty `String`s and `Vec`s don't allocate at all which means they are fine to use as an empty return.